### ISSUE TITLE
x/tools/go/packages: confusing error when loading source newer than toolchain used to build application

### ISSUE THREAD

**Issue by adonovan** [golang/go#55883]
When an application that links go/packages is built with a go1.18 toolchain, it includes go1.18's type checker (go/types). If the application is then used with version go1.20 of the go command on the PATH, go list will report source files that may rely on language features of go1.20 (such as recent additions to the unsafe package), which aren't supported by its type checker. The errors are confusing; see https://github.com/golang/go/issues/55045.

Instead, go/packages should either instruct go list to select files using release tags corresponding to the version of the type checker linked with it, or should otherwise detect and report the version skew in a straightforward message.

**Comment by heschi** [comment]
It seems strange to issue a warning without knowing what the caller is going to do with the results of the call. What if they only want to list the files? Returning a hard error if `NeedTypes` is specified makes more sense to me, FWIW.

**Comment by adonovan** [comment]
Yes, the same thought occurred to me, thought I'm still not sure a hard error is warranted. What if someone updated their go toolchain but it still only using it to build older code? It seems rough that they would have to rebuild any apps that use go/packages. Perhaps we should continue to issue a warning but only if the parser is used. (The problem is broader than NeedTypes.)
